build failure with gcc 14.2.1 on Fedora 41 x86_64 at 0f8969e and related telemetry concerns

N.B. I can see how the update checker would be helpful to some users. However, many others do _not_ want to make such connections, or have any telemetry that cannot be disabled, especially after the earlier fiascos with EA and GameSpy. So, please, if you are going to include such functionality, provide a way for users to easily disable it via cvar and configuration menus before it is used, and mention this prominently in the documents.

Linking 

[openmohaa_0f8969e_build_failure_gcc_14.2.1_fedora41_x86_64_20250312.txt](https://github.com/user-attachments/files/19218269/openmohaa_0f8969e_build_failure_gcc_14.2.1_fedora41_x86_64_20250312.txt)

 failure:

...
[709/716] Linking CXX executable openmohaa.x86_64
FAILED: openmohaa.x86_64 
: && /usr/bin/g++  -Wl,--dependency-file,CMakeFiles/openmohaa.dir/link.d CMakeFiles/openmohaa.dir/misc/dummy.c.o CMakeFiles/openmohaa.dir/code/sys/con_log.c.o CMakeFiles/openmohaa.dir/code/sys/sys_autoupdater.c.o CMakeFiles/openmohaa.dir/code/sys/sys_main.c.o CMakeFiles/openmohaa.dir/code/sys/new/sys_main_new.c.o CMakeFiles/openmohaa.dir/code/sys/new/sys_unix_new.c.o CMakeFiles/openmohaa.dir/code/sys/con_tty.c.o CMakeFiles/openmohaa.dir/code/sys/sys_unix.c.o CMakeFiles/openmohaa.dir/code/sys/win_bounds.cpp.o CMakeFiles/openmohaa.dir/code/sys/win_localization.cpp.o CMakeFiles/openmohaa.dir/code/sys/sys_update_checker.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/puff.c.o CMakeFiles/openmohaa.dir/code/qcommon/q_math.c.o CMakeFiles/openmohaa.dir/code/qcommon/q_shared.c.o CMakeFiles/openmohaa.dir/code/qcommon/class.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/con_set.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/con_timer.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/delegate.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/lightclass.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/listener.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/lz77.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/mem_blockalloc.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/mem_tempalloc.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/script.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/str.cpp.o CMakeFiles/openmohaa.dir/code/script/scriptexception.cpp.o CMakeFiles/openmohaa.dir/code/script/scriptvariable.cpp.o CMakeFiles/openmohaa.dir/code/server/sv_ccmds.c.o CMakeFiles/openmohaa.dir/code/server/sv_client.c.o CMakeFiles/openmohaa.dir/code/server/sv_game.c.o CMakeFiles/openmohaa.dir/code/server/sv_init.c.o CMakeFiles/openmohaa.dir/code/server/sv_main.c.o CMakeFiles/openmohaa.dir/code/server/sv_net_chan.c.o CMakeFiles/openmohaa.dir/code/server/sv_snapshot.c.o CMakeFiles/openmohaa.dir/code/server/sv_snd.c.o CMakeFiles/openmohaa.dir/code/server/sv_world.c.o CMakeFiles/openmohaa.dir/code/qcommon/alias.c.o CMakeFiles/openmohaa.dir/code/qcommon/bg_compat.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/cm_fencemask.c.o CMakeFiles/openmohaa.dir/code/qcommon/cm_load.c.o CMakeFiles/openmohaa.dir/code/qcommon/cm_patch.c.o CMakeFiles/openmohaa.dir/code/qcommon/cm_polylib.c.o CMakeFiles/openmohaa.dir/code/qcommon/cm_terrain.c.o CMakeFiles/openmohaa.dir/code/qcommon/cm_test.c.o CMakeFiles/openmohaa.dir/code/qcommon/cm_trace.c.o CMakeFiles/openmohaa.dir/code/qcommon/cm_trace_lbd.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/cm_trace_obfuscation.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/cmd.c.o CMakeFiles/openmohaa.dir/code/qcommon/common.c.o CMakeFiles/openmohaa.dir/code/qcommon/crc.c.o CMakeFiles/openmohaa.dir/code/qcommon/cvar.c.o CMakeFiles/openmohaa.dir/code/qcommon/files.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/ioapi.c.o CMakeFiles/openmohaa.dir/code/qcommon/huffman.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/md4.c.o CMakeFiles/openmohaa.dir/code/qcommon/md5.c.o CMakeFiles/openmohaa.dir/code/qcommon/memory.c.o CMakeFiles/openmohaa.dir/code/qcommon/msg.cpp.o CMakeFiles/openmohaa.dir/code/qcommon/net_chan.c.o CMakeFiles/openmohaa.dir/code/qcommon/net_ip.c.o CMakeFiles/openmohaa.dir/code/qcommon/unzip.c.o CMakeFiles/openmohaa.dir/code/gamespy/sv_gamespy.c.o CMakeFiles/openmohaa.dir/code/gamespy/sv_gqueryreporting.c.o CMakeFiles/openmohaa.dir/code/zlib/adler32.c.o CMakeFiles/openmohaa.dir/code/zlib/crc32.c.o CMakeFiles/openmohaa.dir/code/zlib/inffast.c.o CMakeFiles/openmohaa.dir/code/zlib/inflate.c.o CMakeFiles/openmohaa.dir/code/zlib/inftrees.c.o CMakeFiles/openmohaa.dir/code/zlib/zutil.c.o CMakeFiles/openmohaa.dir/code/client/cl_avi.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_cgame.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_cin.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_consolecmds.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_curl.c.o CMakeFiles/openmohaa.dir/code/client/cl_input.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_inv.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_invrender.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_keys.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_main.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_net_chan.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_parse.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_scrn.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_ui.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uibind.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uidmbox.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uifilepicker.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uigamespy.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uigmbox.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uilangame.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uiloadsave.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uimaprotationsetup.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uimaprunner.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uiminicon.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uimpmappicker.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uiplayermodelpicker.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uiradar.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uiserverlist.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uisoundpicker.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uistd.cpp.o CMakeFiles/openmohaa.dir/code/client/cl_uiview3d.cpp.o CMakeFiles/openmohaa.dir/code/client/libmumblelink.c.o CMakeFiles/openmohaa.dir/code/client/qal.c.o CMakeFiles/openmohaa.dir/code/client/snd_dma_new.cpp.o CMakeFiles/openmohaa.dir/code/client/snd_info.cpp.o CMakeFiles/openmohaa.dir/code/client/snd_mem_new.cpp.o CMakeFiles/openmohaa.dir/code/client/snd_miles_new.cpp.o CMakeFiles/openmohaa.dir/code/client/snd_openal_new.cpp.o CMakeFiles/openmohaa.dir/code/client/usignal.cpp.o CMakeFiles/openmohaa.dir/code/client/snd_codec.c.o CMakeFiles/openmohaa.dir/code/client/snd_codec_mp3.c.o CMakeFiles/openmohaa.dir/code/client/snd_codec_ogg.c.o CMakeFiles/openmohaa.dir/code/client/snd_codec_opus.c.o CMakeFiles/openmohaa.dir/code/client/snd_codec_wav.c.o CMakeFiles/openmohaa.dir/code/uilib/ucolor.cpp.o CMakeFiles/openmohaa.dir/code/uilib/ui_init.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uibind.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uibindlist.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uibutton.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uicheckbox.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uiconsole.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uidialog.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uifield.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uifloatwnd.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uifont.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uiglobalgamelist.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uihorizscroll.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uilabel.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uilangamelist.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uilayout.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uilist.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uilistbox.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uilistctrl.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uimenu.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uimessage.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uimledit.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uinotepad.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uipopupmenu.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uipulldownmenu.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uipulldownmenucontainer.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uislider.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uistatus.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uivertscroll.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uiwidget.cpp.o CMakeFiles/openmohaa.dir/code/uilib/uiwinman.cpp.o -o openmohaa.x86_64  code/renderercommon/libjpeg8.a  -ldl  -lm  -lrt  -lstdc++  code/gamespy/libgcd_key.a  code/gamespy/libgcd_qr2.a  code/gamespy/libgcd_natneg.a  code/gamespy/libgcd_common.a  code/sdl/libomohsdl_client.a  code/qcommon/tiki/libomohtiki.a  code/qcommon/skeletor/libomohskeletor.a  code/client/libmad/libmad.a  code/renderercommon/renderergl1/renderer_opengl1.x86_64.a  code/renderercommon/libomohrenderer_common.a  code/renderercommon/libjpeg8.a  code/sdl/libomohsdl_gl.a  /usr/lib64/libSDL2main.a  /usr/lib64/libSDL2-2.0.so.0.3000.11 && :
/usr/bin/ld: CMakeFiles/openmohaa.dir/code/client/cl_main.cpp.o: in function `CL_VerifyUpdate()':
cl_main.cpp:(.text+0x4731): undefined reference to `updateChecker'
collect2: error: ld returned 1 exit status
[712/716] Linking CXX shared library code/server/fgame/game.x86_64.so
ninja: build stopped: subcommand failed.

--- a/code/sys/sys_update_checker.cpp
+++ b/code/sys/sys_update_checker.cpp
@@ -56,6 +56,7 @@ UpdateChecker::UpdateChecker()
     versionChecked                    = false;
     handle                            = NULL;
     thread                            = NULL;
+    requestThreadIsActive             = qfalse;
 }
 
 UpdateChecker::~UpdateChecker()
@@ -89,13 +90,39 @@ void UpdateChecker::Init()
     }
 
     curl_easy_setopt(handle, CURLOPT_USERAGENT, "curl");
+#endif
 
-    thread = new std::thread(&UpdateChecker::RequestThread, this);
+    CheckInitClientThread();
+}
+
+void UpdateChecker::CheckInitClientThread()
+{
+    if (!requestThreadIsActive && CanHaveRequestThread()) {
+        requestThreadIsActive = qtrue;
+
+        thread = new std::thread(&UpdateChecker::RequestThread, this);
+    }
+}
+
+bool UpdateChecker::CanHaveRequestThread() const
+{
+    if (!Cvar_VariableIntegerValue("net_enabled")) {
+        // Network has been disabled by a cvar
+        return false;
+    }
+
+#ifdef HAS_LIBCURL
+    return true;
+#else
+    return false;
 #endif
 }
 
 void UpdateChecker::Process()
 {
+    // Initialize the client thread when necessary
+    CheckInitClientThread();
+
     std::chrono::time_point<std::chrono::steady_clock> currentTime = std::chrono::steady_clock::now();
     if (currentTime
         < lastMessageTime + std::chrono::milliseconds(Q_max(1, com_updateCheckInterval->integer) * 60 * 1000)) {
@@ -233,7 +260,6 @@ bool UpdateChecker::ParseVersionNumber(const char *value, int& major, int& minor
     return true;
 }
 
-#ifdef HAS_LIBCURL
 size_t WriteCallback(char *contents, size_t size, size_t nmemb, void *userp)
 {
     std::string& responseString = *(std::string *)userp;
@@ -244,6 +270,7 @@ size_t WriteCallback(char *contents, size_t size, size_t nmemb, void *userp)
 
 void UpdateChecker::DoRequest()
 {
+#ifdef HAS_LIBCURL
     std::lock_guard<std::shared_mutex> l(clientMutex);
     CURLcode                           result;
     std::string                        responseString;
@@ -276,6 +303,7 @@ void UpdateChecker::DoRequest()
 
         versionChecked = true;
     } catch (std::out_of_range&) {}
+#endif
 }
 
 void UpdateChecker::RequestThread()
@@ -283,7 +311,7 @@ void UpdateChecker::RequestThread()
     std::chrono::time_point<std::chrono::steady_clock> currentTime = std::chrono::steady_clock::now();
     std::chrono::time_point<std::chrono::steady_clock> lastCheckTime;
 
-    while (handle) {
+    while (handle && CanHaveRequestThread()) {
         currentTime = std::chrono::steady_clock::now();
         if (currentTime
             >= lastCheckTime + std::chrono::milliseconds(Q_max(1, com_updateCheckInterval->integer) * 60 * 1000)) {
@@ -294,5 +322,6 @@ void UpdateChecker::RequestThread()
 
         std::this_thread::sleep_for(std::chrono::seconds(5));
     }
+
+    requestThreadIsActive = qfalse;
 }
-#endif

--- a/code/sys/sys_update_checker.cpp
+++ b/code/sys/sys_update_checker.cpp
@@ -24,14 +24,14 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 #include "../qcommon/q_version.h"
 
 #ifdef HAS_LIBCURL
-
 #    include <curl/curl.h>
+#endif
 
 //#include <httplib.h>
-#    include "../qcommon/json.hpp"
+#include "../qcommon/json.hpp"
 using json = nlohmann::json;
 
-#    include <chrono>
+#include <chrono>
 
 UpdateChecker updateChecker;
 
@@ -67,6 +67,7 @@ UpdateChecker::~UpdateChecker()
 
 void UpdateChecker::Init()
 {
+#ifdef HAS_LIBCURL
     CURLcode result;
 
     assert(!handle);
@@ -90,6 +91,7 @@ void UpdateChecker::Init()
     curl_easy_setopt(handle, CURLOPT_USERAGENT, "curl");
 
     thread = new std::thread(&UpdateChecker::RequestThread, this);
+#endif
 }
 
 void UpdateChecker::Process()
@@ -122,6 +124,7 @@ void UpdateChecker::Shutdown()
 
 void UpdateChecker::ShutdownClient()
 {
+#ifdef HAS_LIBCURL
     std::lock_guard<std::shared_mutex> l(clientMutex);
 
     if (!handle) {
@@ -130,6 +133,7 @@ void UpdateChecker::ShutdownClient()
 
     curl_easy_cleanup(handle);
     handle = NULL;
+#endif
 }
 
 void UpdateChecker::ShutdownThread()
@@ -229,6 +233,7 @@ bool UpdateChecker::ParseVersionNumber(const char *value, int& major, int& minor
     return true;
 }
 
+#ifdef HAS_LIBCURL
 size_t WriteCallback(char *contents, size_t size, size_t nmemb, void *userp)
 {
     std::string& responseString = *(std::string *)userp;
@@ -290,23 +295,4 @@ void UpdateChecker::RequestThread()
         std::this_thread::sleep_for(std::chrono::seconds(5));
     }
 }
-
-#else
-
-bool UpdateChecker::CheckNewVersion() const
-{
-    return false;
-}
-
-bool UpdateChecker::CheckNewVersion(int& major, int& minor, int& patch) const
-{
-    return false;
-}
-
-void Sys_UpdateChecker_Init() {}
-
-void Sys_UpdateChecker_Process() {}
-
-void Sys_UpdateChecker_Shutdown() {}
-
 #endif

--- a/code/sys/sys_update_checker.h
+++ b/code/sys/sys_update_checker.h
@@ -59,6 +59,8 @@ private:
     void ShutdownClient();
     void ShutdownThread();
     void RequestThread();
+    void CheckInitClientThread();
+    bool CanHaveRequestThread() const;
     void DoRequest();
     bool ParseVersionNumber(const char *value, int& major, int& minor, int& patch) const;
 
@@ -79,6 +81,7 @@ private:
     void             *handle;
     std::shared_mutex clientMutex;
     std::thread      *thread;
+    qboolean          requestThreadIsActive;
 };
 
 extern UpdateChecker updateChecker;

